### PR TITLE
Wrote tests for DownloadTaskProgress

### PR DESCRIPTION
**Pull request by singhraman4282**
**What's this do?**
Adds tests for NYPLAxisDownloadProgress

**Why are we doing this? (w/ JIRA link if applicable)**
Testing purposes

**How should this be tested? / Do these changes have associated tests?**
Cannot be QA tested

**Dependencies for merging? Releasing to production?**
N/A

**Does this include changes that require a new SimplyE build for QA?**
N/A

**Has the application documentation been updated for these changes?**
N/A

**Did someone actually run this code to verify it works?**
Raman Singh verified it

### CODE REVIEW

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: licenseURL), 0.05)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: containerURL), 0.05)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: encryptionURL), 0.05)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.85/3)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter2URL), 0.85/3)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter3URL), 0.85/3)
+
+    // ~43% completed
+    sut.didFinishTask(with: chapter2URL)
+    XCTAssertEqual(progressListener.currentProgress, (0.15 + (0.85/3)).roundedToTwoDecimalPlaces)
```
Comment (ettore) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:166]: great test -- but if you round it why not just using `0.43`?

File: SimplifiedTests/NYPLAxisTests/NYPLAxisItemDownloaderTests.swift [hunk 2]
```diff
@@ -198,4 +203,71 @@ class NYPLAxisItemDownloaderTests: XCTestCase {
...
+    
+    let itemDownloader = NYPLAxisItemDownloader(
+      assetWriter: assetWriter, dispatchGroup: dispatchGroup,
+      downloader: contentDownloader, weightProvider: weightProvider)
+    
+    let progressListener = NYPLAxisProgressListenerMock()
+    itemDownloader.delegate = progressListener
+    XCTAssertEqual(progressListener.currentProgress, 0)
+    
+    dispatchGroup.enter()
```
Comment (ettore) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisItemDownloaderTests.swift:218]: would you say this usage of the dispatchGroup assumes knowledge of how the group is used inside the  code being tested?  Or is this documented? it probably is/should, but I don't recall. From this point of view I wouldn't know why I need to "enter" the group here and wait later, for example. If it's already documented in NYPLAxisItemDownloader, this is fine.

**Review by ettore — APPROVED** [review]

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.45)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter2URL), 0.45)
+
+    sut.addFlexibleWeightTask(with: encryptionURL)
+    // 45 % downloaded
+    XCTAssertEqual(progressListener.currentProgress, 0.45)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: licenseURL), 0.05)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: containerURL), 0.05)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.45)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter2URL), 0.45)
```
Comment (ErnestFan) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:98]: Correct me if I am wrong, we now have two unfinished flexible weight tasks, should the weight of `chapter2URL` be 0.225 ((1 - 0.05 - 0.05 - 0.45) / 2)?

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.8)
+
+    sut.addFlexibleWeightTask(with: chapter3URL)
+    XCTAssertEqual(progressListener.currentProgress, 0.0)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: licenseURL), 0.2)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.4)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter3URL), 0.4)
+
+    // 20 % completed
+    sut.didFinishTask(with: licenseURL)
```
Comment (ErnestFan) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:190]: This test is named `AddingMoreTasksShouldNotAffectWeightOfCompletedTasks` but there is no task added after this task is completed. Also `licenseURL` is a fixed weight task, so it should not be affected in any ways?

File: SimplifiedTests/NYPLAxisTests/NYPLAxisItemDownloaderTests.swift (hunk 2 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisItemDownloaderTests.swift:218]: You're right. We're entering dispatchGroup from multiple places. However, NYPLAxisItemDownloader should be the only place where we enter and leave the dispatchGroup and other classes should not be aware of it. Added a TODO note in code. Will work on that.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift (hunk 3 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift]: You sir are absolutely right. Not only was there a problem with my math, but there was also a bug 🤮  in my logic. Thank you for pointing it out. Fixed.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift (hunk 4 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift]: Great catch. Fixed.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift (hunk 1 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:167]: You're right. It is 43%.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  func testAddingMoreTasksShouldNotAffectWeightOfCompletedTasks() {
+    sut.addFixedWeightTask(with: licenseURL, weight: 0.2)
+    sut.addFlexibleWeightTask(with: chapter1URL)
+
+    XCTAssertEqual(progressListener.currentProgress, 0.0)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: licenseURL), 0.2)
+    XCTAssertEqual(sut.test_getWeightForProcess(withURL: chapter1URL), 0.8)
+    
+    // 20 % completed
+    sut.didFinishTask(with: licenseURL)
```
Comment (ErnestFan) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:185]: We are checking the weight of a fixed weight task here, shouldn't we test a flexible weight task instead?

Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisDownloadProgressTests.swift:185]: done

**Review by ErnestFan — APPROVED** [review]

**Pull request merged**